src/activatable/activatable.ts

```diff
@@ -0,0 +1,130 @@
+/**
+ * Created ->- Activating ->- Activated ->- Deactivating ->- Deactivated -------------->----------- Disposed
+ *                 |  |                                         |  |                                  |  |
+ *                 |  `------------------<----------------------'  `--->--- FailedToDeactivate --->---'  |
+ *                 `--------------------->--------- FailedToActivate ------------------>-----------------'
+ */
+
+enum LifeCycleState {
+  Created = 1,
+  Activating,
+  Activated,
+  Deactivating,
+  Deactivated,
+  Disposed,
+  FailedToActivate,
+  FailedToDeactivate,
+}
+
+const LifeCycleDescription = {
+  [LifeCycleState.Created]: 'has never activated before',
+  [LifeCycleState.Activating]: 'is activating',
+  [LifeCycleState.Activated]: 'is activated',
+  [LifeCycleState.Deactivating]: 'is deactivating',
+  [LifeCycleState.Deactivated]: 'is deactivated',
+  [LifeCycleState.Disposed]: 'is disposed',
+  [LifeCycleState.FailedToActivate]: 'errored while activating',
+  [LifeCycleState.FailedToDeactivate]: 'errored while deactivating',
+};
+
+export interface IActivatable {
+  isActivating: boolean;
```

Comment:
Let's make state public and remove all these `isSomething` members.

Reply:
keeping them will help us migrate all our existing code that relies on these properties.

Reply:
These properties are annoying, frankly speaking. It's usually easier to control the flow if you have a single property you can rely on.
See, you don't use them in your own code.
You prefer to use `switch` statement than these properties. :laughing:

Reply:
yeah, but they are so convenient as guards, which is 90% of what we use them for. I think I'll keep them, but add the missing ones and keep them dead simple so they correspond exactly.